pkg/sentry/fsimpl/fuse/dev.go

```diff
@@ -79,7 +120,73 @@ func (fd *DeviceFD) Read(ctx context.Context, dst usermem.IOSequence, opts vfs.R
 		return 0, syserror.EPERM
 	}
 
-	return 0, syserror.ENOSYS
+	kernelTask := kernel.TaskFromContext(ctx)
+	if kernelTask == nil {
+		log.Warningf("fusefs.DeviceFD.Read: couldn't get kernel task from context")
+		return 0, syserror.EINVAL
+	}
+
+	// Wait for a request to be made available.
+	if err := kernelTask.Block(fd.emptyQueueCh); err != nil {
+		log.Warningf("fusefs.DeviceFD.Read: couldn't wait on request queue: %v", err)
+		return 0, syserror.EBUSY
+	}
+
+	fd.mu.Lock()
+	defer fd.mu.Unlock()
+	return fd.readLocked(ctx, dst, opts)
+}
+
+// readLocked implements the reading of the fuse device while locked with DeviceFD.mu.
+func (fd *DeviceFD) readLocked(ctx context.Context, dst usermem.IOSequence, opts vfs.ReadOptions) (int64, error) {
+	if fd.queue.Empty() {
+		log.Warningf("fusefs.DeviceFD.Read: No requests to read but still signalled")
+		return 0, syserror.EAGAIN
+	}
+
+	req := fd.queue.Front()
+	if fd.readCursor >= req.hdr.Len {
+		// Cursor points past end of current request payload? Reset the cursor,
+		// remove the front request and try again.
+		fd.readCursor = 0
+		fd.queue.Remove(req)
+		fd.numInFlightRequests -= 1
+
+		// Signal that the queue is no longer full.
+		select {
+		case fd.fullQueueCh <- struct{}{}:
```

Comment:
but the request is not complete yet, Should we signal when we get a response for this request?